### ISSUE TITLE
Mr. Doc Next

### ISSUE THREAD

**Issue by iwatakeshi** [mr-doc/mr-doc#94]
@FGRibreau 

I don't plan to start working on this any time soon but I am definitely interested in making a blueprint for the next Mr. Doc. This will definitely require some help because I'm pretty much limited in the following field. To begin, I'll just put my thoughts on the table and see where we can go from here.

My thoughts:

* Create a parser that is language agnostic and pluggable.
* As mentioned in #93, I think Mr. Doc depends too much on dox.
  * dox [isn't very active](https://github.com/tj/dox/pulse/monthly)
  * dox's [documentation sucks](https://github.com/tj/dox#readme) (oh the irony)
  * Personally, I don't understand how dox works or even how parsing works in general but [it's been in my interest in creating one](https://github.com/iwatakeshi/Metal). So far I somewhat understand how lexing/scanning works but I haven't had time to look into parsing so if we create one for Mr. Doc, it will be a plus for me.

* Create our own standard which will be a subset of JSDoc and Closure compiler
* Add additional template compilers besides Jade. This will bring more web developers to enhance the themes.

*The "Next" just means "the next version of Mr. Doc" which could be 4.0.0 or 5.0.0 depending on what happens after 3.0.0.*

**Comment by FGRibreau** [comment]
Indeed, if we create an intermediate language (just like LLVM does) we will be able to plug any language parser to it and still use the same frontend :)

**Comment by iwatakeshi** [comment]
:+1: Yeah, LLVM might be something to look into. I'll be going over their [tutorial](http://llvm.org/docs/tutorial/#kaleidoscope-implementing-a-language-with-llvm).
Would you have any thoughts of your own on how we could implement this? :)

**Comment by gr2m** [comment]
Thanks for all your efforts! I think automated documentation from source is one of these things that don't have a good solution yet. We use `doxx` all over, but are not happy with, so I'm glad you address the problems. I'm happy to help reach out to find more contributors / help creating the new parser. I'm pretty sure many people have the same struggles, and moving away from dox sounds like a good idea at this point

**Comment by iwatakeshi** [comment]
@gr2m Thanks for the feedback! I would really appreciate any help! : ) As mentioned above, this is beyond my realm but would love to learn. It would be great to have a few contributors to put some thought into this so that we can come up with a blueprint/spec that outlines how to build a language agnostic documenter that isn't completely complex and extendable (such as in the languages to support and the themes). Let me know if there is anything I can do for you ;)

**Comment by gr2m** [comment]
I'd put the parsers into their own npm modules. I'd probably start `mr-doc` with a parser for JavaScript only and make it work, then think about a plugabble API as a second step. 

From my experience, I wouldn't try to come up with a full blueprint/spec and make it fully language agnostic, instead I'd think of what a first useful mr-doc version would look like which aligns with that goal in the long term, then build the first version and build upon it

**Comment by iwatakeshi** [comment]
Good point. We'll have to create a repo for it...maybe name it `core` and have the tokenizer and parser in the same repo but under separate folders?

JS will definitely be the focus. I don't plan to build a plugin for other languages but will leave it to others who want to extend the core. We'll have to make sure that we don't mix things in the core that is JS specific.

**Comment by iwatakeshi** [comment]
@gr2m Any thoughts?

**Comment by iwatakeshi** [comment]
@gr2m In the meantime, I just wrote a small gist that shows a basic usage of the scanner.
https://gist.github.com/iwatakeshi/d1973038c1417443749f

**Comment by FGRibreau** [comment]
@iwatakeshi If we go this way, we should use for http://esprima.org/ and then convert the AST into our own intermediate language (IL) to support other languages as well.

**Comment by gr2m** [comment]
I'm super sorry @iwatakeshi, I'm still catching up with a trillion things right now after vacation, but I did not forget about you. I need to give this some more thoughts, I'll get back to you next week, pinky promise

**Comment by iwatakeshi** [comment]
@FGRibreau I've looked into esprima but it's Javascript specific correct? I wouldn't mind using esprima or acorn as a guide line for building an AST though. I prefer the way babel outputs the AST (see http://astexplorer.net/ and change the parser to babylon) because it includes the comments. :)

I do agree that we could just use esprima but again I think that is Javascript specific and I would like to extend beyond Javascript by using plugins. Its true that its a hard project, but if we do it as a team, keep things simple, and take our time then I'm sure we can get it right without worrying about difficulty. ;)

@gr2m No worries, this project isn't in a big rush. I've just released 3.0.0 so it will be a long ways until we release the next big major update. ;)

@FGRibreau @gr2m Check out the [gist](https://gist.github.com/iwatakeshi/d1973038c1417443749f) and try it with `babel-node` and let me know what you think.

**Comment by iwatakeshi** [comment]
@FGRibreau @gr2m This is a rough idea but you can check out a repo I made which is more organized that the gist file:
https://github.com/iwatakeshi/cherry
If we do go this route then I will just transfer the repo into mr-doc but if not then I'll just leave at is. ;)

**Comment by gr2m** [comment]
@iwatakeshi it feels a bit like you try to achieve too much at once. `mr-doc` is JavaScript-specific right now, and at least for me, that is a feature, not a problem. I think `esprima` widely used, if you get it going quickly, why not go this path, make mr-doc useful with it, and worry about other languages later, if at all.

From a user's perspective, I care more about how I can configure own templates / own template formats, and how I can extend the parser with own `@xxx` tags, for example.

**Comment by iwatakeshi** [comment]
@gr2m *sigh* Well I did not plan to achieve anything at once but I guess we could go the esprima route. Although I prefer the way babylon output's the AST (see http://astexplorer.net/).

**Comment by iwatakeshi** [comment]
@gr2m I've add you as a collaborator :). Now it's just figuring out how we can allow users to extend the AST and the templates.

**Comment by iwatakeshi** [comment]
@FGRibreau @gr2m One thing that comes to mind is using esprima (or some parser x) to parse the source and then create another parser for the comments...similar to [doctrine](https://github.com/eslint/doctrine) but with a capability to extend.

**Comment by iwatakeshi** [comment]
@FGRibreau  @gr2m Well since there hasn't been much activity, I decided to continue to work on [cherry](https://github.com/iwatakeshi/cherry) since it's quite powerful as a generic scanner and parser. The scanner works really well and the format looks exactly like Babylon's output (the token part). I haven't started on the parser because I would like some advice in that area. For inspiration, we can look at [VSCode's source code](https://github.com/Microsoft/vscode/tree/master/src/vs/languages/css/common/parser) and apply it toward cherry. Anyways, any help will be greatly appreciated.

**Comment by gr2m** [comment]
@iwatakeshi I'm sorry I didn't respond, it’s not that I'm not interested, just overworked right now, I've to watch out. But I have it on my list of things, and as soon as I find some time, I'll look into your progress and give feedback

**Comment by iwatakeshi** [comment]
@gr2m :+1:

**Comment by iwatakeshi** [comment]
A couple of resources I am using for the parser: 

* http://journal.stuffwithstuff.com/2011/03/19/pratt-parsers-expression-parsing-made-easy/
* http://www.codeproject.com/Articles/345888/How-to-write-a-simple-interpreter-in-JavaScript

**Comment by iwatakeshi** [comment]
Just wanted to leave some progress update and more thoughts to give us some direction:

Through out this week, I've been working on [cherry](https://github.com/iwatakeshi/cherry) and studying about compilers and I have to say, I have learned a lot. The amount I've learned is probably not close enough to fully understand how acorn or esprima works but it was enough to use cherry to create a simple language called "ac" (Adding Calculator) from [Crafting a Compiler by Charles N. Fischer](http://www.amazon.com/Crafting-Compiler-Charles-N-Fischer/dp/0136067050). 

There is still more reading/work to be done to implement a complete parser and AST but a working example can be found [on rawgit](https://rawgit.com/iwatakeshi/cherry/master/html/index.html). Make sure to use Chrome or Edge because I'm using ES6 keywords such as `let` or `const`.

Another thing I would like to mention is about documentation style. As I was working with cherry, I felt as if JSDoc was lacking something. I did mention in the issue about creating a subset of JSDoc and so far the following was the syntax I used to [document cherry](https://github.com/iwatakeshi/cherry/blob/master/lib/Scanner.js):

**Documenting a class:**
```javascript
/**
  @class {Car} - A car class that can do awesome things.
  @constructor {maker: string} - The maker of the car.
  @example: javascript  {
     const car = new Car('Honda');
     // ...
  }
*/
class Car {
  constructor (maker)  {
      //...
  }
} 
```

**Documenting a method:**
```javascript
/**
  @method {greet} - Greets the user.
  @param {name?:string} - The user's name.
*/
function greet (name) {
  console.log('Hello, ' + name ? name : 'user!');
}
```

I will admit that some ideas came from TypeScript. I really like TypeScript because it gives the code more meaning when it comes to types and I think that's how documenting should work. i.e. `?` can tell me that something is optional or `name?:string` can tell me that the name should be a string but is optional.

**Comment by iwatakeshi** [comment]
Just wanted give an update on the progress here:

I've decided to release [tomo](https://github.com/iwatakeshi/tomo) (formerly known as cherry) as an npm package so that anyone can start contributing. Anyone reading this progress update are free to contribute to the project.

**Comment by iwatakeshi** [comment]
@FGRibreau @gr2m Hey guys, if you know any developer who has time to contribute please tell them about this awesome project! I've started working on [4.0.0](https://github.com/mr-doc/mr-doc/tree/4.0.0) and would love to get some help. Unfortunately, I won't be using my generic parser but that can be added in the future thanks to this upcoming major update. 

The project isn't going to be completed any time soon and I'm projecting that it might be done by August. Of course, it will depend on the amount of contribution that Mr. Doc gets.

**Comment by gr2m** [comment]
@iwatakeshi check out initiatives like http://yourfirstpr.github.io/ and others mentioned here: http://firsttimersonly.com/. It takes a bit invest to prepare issues for new contributors, but it works to grow the contributors base over time, and it’s a lot of fun. 

I recently wrote about how we do starter issues at Hoodie: http://hood.ie/blog/starter-issues.html

I’m sorry I can’t help better myself, but Hoodie keeps me fully occupied right now, as we closing on a release. Thanks for your continuous work on this, I think the JS ecosystem really needs a good docs solution, and Mr. Doc looks like it could become it

**Comment by iwatakeshi** [comment]
@gr2m Thanks for the resources and not a problem! I understand that we are all occupied. :)

**Comment by iwatakeshi** [comment]
**Update:**

Since 2016 I have been writing and re-writing (or more like trying and failing 🤣 ) but I've finally got some idea of where I'd like to go:

* The way documentation is written will change. Every time I looked at JSDoc's documentation page, it was always vague when it came to describing a certain data type. For example, writing `string[]` is much easier than `Array.<string>`. I've written the syntax parser and it's available [on my Github page](www.github.com/iwatakeshi/xdoc-parser).

* mr-doc will hopefully be something like Rust's documentation tool. It was when I had to use Rust to demonstrate its parallelization capabilities in class that I discovered its documentation tool. Long story short, it was amazing and I wished other languages had a tool like that built in.

* Becoming a language agnostic documentation generator is still a goal and it may be possible with [tree-sitter](www.github.com/tree-sitter). I have a pull request open for suggestions at https://github.com/mr-doc/mr-doc-parser/pull/11. cc @FGRibreau 

* Theming is still a goal but probably won't land in v4. I'd like some suggestions on how to go about this. I've done some preliminary research and I looked into `parcel.js`, `react.js`and `vue.js`. What I would like to do is have `mr-doc-compiler` dynamically generate a template (or components) and then we can have users provide a css file to modify each component by class. i.e. `.navbar` etc.
